src/runtime/objmodel.cpp

```diff
@@ -3914,7 +3908,93 @@ extern "C" Box* unaryop(Box* operand, int op_type) {
     return rtn;
 }
 
-extern "C" Box* getitem(Box* value, Box* slice) {
+Box* callItemAttr(Box* target, const std::string& item_str, Box* item, Box* value, CallRewriteArgs* rewrite_args) {
+    if (value) {
+        return callattrInternal2(target, item_str, CLASS_ONLY, rewrite_args, ArgPassSpec(2), item, value);
+    } else {
+        return callattrInternal1(target, item_str, CLASS_ONLY, rewrite_args, ArgPassSpec(1), item);
+    }
+}
+
+// This function decides whether to call the slice operator (e.g. __getslice__)
+// or the item operator (__getitem__).
+Box* callItemOrSliceAttr(Box* target, const std::string& item_str, const std::string& slice_str, Box* slice, Box* value,
+                         CallRewriteArgs* rewrite_args) {
+    // Guard on the type of the object (want to have the slice operator attribute).
+    Box* slice_attr = NULL;
+    if (rewrite_args) {
+        RewriterVar* target_cls = rewrite_args->obj->getAttr(offsetof(Box, cls));
+        GetattrRewriteArgs grewrite_args(rewrite_args->rewriter, target_cls, Location::any());
+        slice_attr = typeLookup(target->cls, slice_str, &grewrite_args);
+        if (!grewrite_args.out_success) {
+            rewrite_args = NULL;
+        }
+    } else {
+        slice_attr = typeLookup(target->cls, slice_str, NULL);
+    }
+
+    if (!slice_attr) {
+        return callItemAttr(target, item_str, slice, value, rewrite_args);
+    }
+
+    // Need a slice object to use the slice operators.
+    if (slice->cls != slice_cls) {
+        if (rewrite_args) {
+            rewrite_args->arg1->addAttrGuard(offsetof(Box, cls), (uint64_t)slice_cls, /*negate=*/true);
```

Comment:
I would vote for adding guards where we need them, and separately dealing with the resulting duplicate guards (ex peephole optimizer, or remembering what guards we emitted).  ex I think we usually don't need to guard on the arg classes in callattrInternal, but if we fixed that there it would break callers that relied on it doing the guards.